# Model Comparison and Conclusion

This note summarizes the work and compare the models in one sight.

## Libraries

Importing necessary libraries.

In [1]:
import pandas as pd

import numpy as np

from tensorflow.keras.models import load_model

from sklearn.metrics import confusion_matrix

from malig_data import *


train_img shape (5421, 12288)
train_images shape (5421, 64, 64, 3)
train_labels shape (5421, 2)
test_img shape (1355, 12288)
test_images shape (1355, 64, 64, 3)
test_labels shape (1355, 2)
val_img shape (753, 12288)
val_images shape (753, 64, 64, 3)
val_labels shape (753, 2)
smote_img shape (9358, 12288)
smote_images shape (9358, 64, 64, 3)
smote_labels shape (9358,)
adasyn_img shape (9265, 12288)
adasyn_images shape (9265, 64, 64, 3)
adasyn_labels shape (9265,)
train_y shape (5421, 1)
test_y shape (1355, 1)
val_y shape (753, 1)


Defining file Path

In [2]:
basic_NN_file = "../models/basic_NN.h5"
smote_NN_file = "../models/SMOTE_NN.h5"
adasyn_NN_file = "../models/ADASYN_NN.h5"

basic_CNN_file = "../models/basic_cnn.h5"
smote_CNN_file = "../models/smote_cnn.h5"
adasyn_CNN_file = "../models/adasyn_cnn.h5"

VGG19_file = "../models/vgg19.h5"
VGG19_adasyn_file = '../models/vgg19_adasyn.h5'

ResNet50_file = "../models/ResNet50.h5"
ResNet50_adasyn_file = "../models/ResNet50 Adasyn.h5"

In [3]:
def model_loading(filepath, train_X, train_y, val_X, val_y):
    """
    Function that displays the confusion matrix of the model and returns the saved model
    filepath: where the model is saved
    train_X: train image
    train_y: train target
    val_X: validation image
    val_y: validation target
    """
    saved_model = load_model(filepath)

    results_train = saved_model.evaluate(train_X, train_y)
    print(f'Training Loss: {results_train[0]:.3} \nTraining Accuracy: {results_train[1]:.3}')

    print('----------')

    results_test = saved_model.evaluate(val_X, val_y)
    print(f'Test Loss: {results_test[0]:.3} \nTest Accuracy: {results_test[1]:.3}')

    predictions = saved_model.predict_classes(val_X)
    cm = confusion_matrix(val_y, predictions, labels=[0,1])
    
    index=["Actual Malig", "Actual Benign"]
    columns=["Predicted Malig", "Predicted Benign"]
    df = pd.DataFrame(data=cm,index=index, columns=columns)
    display(df)

    return saved_model

Model Comparisons

In [4]:
basic_NN_model = model_loading(basic_NN_file, train_img, train_y, 
                               val_img, val_y)

170/170 [==============================] - 0s 2ms/step - loss: 0.0996 - accuracy: 0.9655
Training Loss: 0.0996 
Training Accuracy: 0.966
----------
24/24 [==============================] - 0s 2ms/step - loss: 0.1975 - accuracy: 0.9363


W0105 23:49:56.012496 4583423424 deprecation.py:323] From <ipython-input-3-eb0f227b5e18>:13: Sequential.predict_classes (from tensorflow.python.keras.engine.sequential) is deprecated and will be removed after 2021-01-01.
Instructions for updating:
Please use instead:* `np.argmax(model.predict(x), axis=-1)`,   if your model does multi-class classification   (e.g. if it uses a `softmax` last-layer activation).* `(model.predict(x) > 0.5).astype("int32")`,   if your model does binary classification   (e.g. if it uses a `sigmoid` last-layer activation).


Test Loss: 0.198 
Test Accuracy: 0.936


Predicted Malig  Predicted Benign
Actual Malig                87                34
Actual Benign               14               618

In [5]:
smote_NN_model = model_loading(smote_NN_file, smote_img, smote_labels,
                               val_img, val_y)

293/293 [==============================] - 3s 9ms/step - loss: 0.0903 - accuracy: 0.9677: 0s - loss: 0.0960 - accuracy:  - ETA: 0s - loss: 0.0946 - accu
Training Loss: 0.0903 
Training Accuracy: 0.968
----------
24/24 [==============================] - 0s 4ms/step - loss: 0.3517 - accuracy: 0.8977
Test Loss: 0.352 
Test Accuracy: 0.898


Predicted Malig  Predicted Benign
Actual Malig               106                15
Actual Benign               62               570

In [6]:
adasyn_NN_model = model_loading(adasyn_NN_file, adasyn_img, adasyn_labels, 
                                val_img, val_y)

290/290 [==============================] - 1s 2ms/step - loss: 0.0915 - accuracy: 0.9662
Training Loss: 0.0915 
Training Accuracy: 0.966
----------
24/24 [==============================] - 0s 3ms/step - loss: 0.4226 - accuracy: 0.8884
Test Loss: 0.423 
Test Accuracy: 0.888


Predicted Malig  Predicted Benign
Actual Malig               100                21
Actual Benign               63               569

In [7]:
basic_CNN_model = model_loading(basic_CNN_file, train_images, train_y, 
                                val_images, val_y)

170/170 [==============================] - 6s 33ms/step - loss: 0.1832 - accuracy: 0.9408
Training Loss: 0.183 
Training Accuracy: 0.941
----------
24/24 [==============================] - 1s 40ms/step - loss: 0.1888 - accuracy: 0.9349
Test Loss: 0.189 
Test Accuracy: 0.935


Predicted Malig  Predicted Benign
Actual Malig                81                40
Actual Benign                9               623

In [8]:
smote_CNN_model = model_loading(smote_CNN_file, smote_images, smote_labels,
                               val_images, val_y)

293/293 [==============================] - 9s 30ms/step - loss: 0.3162 - accuracy: 0.8842
Training Loss: 0.316 
Training Accuracy: 0.884
----------
24/24 [==============================] - 1s 37ms/step - loss: 0.2409 - accuracy: 0.9137
Test Loss: 0.241 
Test Accuracy: 0.914


Predicted Malig  Predicted Benign
Actual Malig                90                31
Actual Benign               34               598

In [9]:
adasy_CNN_model = model_loading(adasyn_CNN_file, adasyn_images, adasyn_labels,
                               val_images, val_y)

290/290 [==============================] - 10s 36ms/step - loss: 0.2567 - accuracy: 0.9005
Training Loss: 0.257 
Training Accuracy: 0.9
----------
24/24 [==============================] - 1s 29ms/step - loss: 0.1913 - accuracy: 0.9456
Test Loss: 0.191 
Test Accuracy: 0.946


Predicted Malig  Predicted Benign
Actual Malig                92                29
Actual Benign               12               620

In [10]:
vgg19_model = model_loading(VGG19_file, train_images, train_y,
                           val_images, val_y)

170/170 [==============================] - 160s 942ms/step - loss: 0.1266 - accuracy: 0.9561
Training Loss: 0.127 
Training Accuracy: 0.956
----------
24/24 [==============================] - 19s 806ms/step - loss: 0.2386 - accuracy: 0.9363
Test Loss: 0.239 
Test Accuracy: 0.936


Predicted Malig  Predicted Benign
Actual Malig                77                44
Actual Benign                4               628

In [11]:
vgg19_model19_adasyn_model = model_loading(VGG19_adasyn_file, adasyn_images, adasyn_labels,
                                   val_images, val_y)

290/290 [==============================] - 257s 888ms/step - loss: 0.0844 - accuracy: 0.9704
Training Loss: 0.0844 
Training Accuracy: 0.97
----------
24/24 [==============================] - 20s 840ms/step - loss: 0.2887 - accuracy: 0.9296
Test Loss: 0.289 
Test Accuracy: 0.93


Predicted Malig  Predicted Benign
Actual Malig                79                42
Actual Benign               11               621

In [12]:
resNet50_model = model_loading(ResNet50_file, train_images, train_y,
                              val_images, val_y)

170/170 [==============================] - 56s 328ms/step - loss: 0.2520 - accuracy: 0.9044
Training Loss: 0.252 
Training Accuracy: 0.904
----------
24/24 [==============================] - 10s 397ms/step - loss: 0.2517 - accuracy: 0.8964
Test Loss: 0.252 
Test Accuracy: 0.896


Predicted Malig  Predicted Benign
Actual Malig                54                67
Actual Benign               11               621

In [13]:
resNet50_adasyn_model = model_loading(ResNet50_adasyn_file, adasyn_images, adasyn_labels,
                                      val_images, val_y)

290/290 [==============================] - 99s 343ms/step - loss: 0.3501 - accuracy: 0.8507
Training Loss: 0.35 
Training Accuracy: 0.851
----------
24/24 [==============================] - 8s 314ms/step - loss: 0.3287 - accuracy: 0.8752
Test Loss: 0.329 
Test Accuracy: 0.875


Predicted Malig  Predicted Benign
Actual Malig                90                31
Actual Benign               63               569

## Conclusion

Based on above observation, the best model to work with is CNN model with adasyn data balance. It has exhibited 94.6% with further tuninng, it would be able to bring higher accuracy.